Tests/TokenizerTests/HTML5LibTestsParser.swift

```diff
@@ -0,0 +1,162 @@
+        self.initialStates = try container.decodeIfPresent([InitialState].self, forKey: .initialStates) ?? [.data]
+        self.lastStartTag = try container.decodeIfPresent(String.self, forKey: .lastStartTag)
+        self.errors = try container.decodeIfPresent([ExpectedError].self, forKey: .errors) ?? []
+    }
+}
+
+struct ExpectedToken {
+    var fields: [ExpectedTokenField?]
+
+    consuming func into() throws -> [Token] {
+        let fields = self.fields
+        switch fields[0] {
+        case .str("DOCTYPE"):
+            guard case (.str(let name), _, _, .bool(let correctness)) = (fields[1], fields[2], fields[3], fields[4]) else {
+                throw TestParseError.invalidTokenFormat
+            }
+            return [.doctype(.init(name: name, forceQuirks: !correctness))]
+        case .str("StartTag"):
+            switch fields.count {
+            case 4:
+                guard case (.str(let name), .dict(let attrs), .bool(true)) = (fields[1], fields[2], fields[3]) else {
+                    throw TestParseError.invalidTokenFormat
+                }
+                return [.tag(.init(name: name, kind: .start, attrs: attrs.map { k, v in .init(name: k, value: v) }, selfClosing: true))]
+            case 3:
+                guard case (.str(let name), .dict(let attrs)) = (fields[1], fields[2]) else {
+                    throw TestParseError.invalidTokenFormat
+                }
+                return [.tag(.init(name: name, kind: .start, attrs: attrs.map { k, v in .init(name: k, value: v) }))]
+            case _: throw TestParseError.invalidTokenFormat
+            }
+        case .str("EndTag"):
+            guard case .str(let name) = fields[1] else { throw TestParseError.invalidTokenFormat }
+            return [.tag(.init(name: name, kind: .end))]
+        case .str("Comment"):
+            guard case .str(let data) = fields[1] else { throw TestParseError.invalidTokenFormat }
+            return [.comment(data)]
+        case .str("Character"):
+            guard case .str(let data) = fields[1] else { throw TestParseError.invalidTokenFormat }
+            return data.map(Token.char)
+        case _: throw TestParseError.invalidTokenType
+        }
+    }
+}
+
+extension ExpectedToken: Decodable {
+    init(from decoder: any Decoder) throws {
+        let container = try decoder.singleValueContainer()
+        self.fields = try container.decode([ExpectedTokenField?].self)
+    }
+}
+
+extension Token {
+}
+
+enum ExpectedTokenField {
+    case str(String)
+    case bool(Bool)
+    case dict([String: String])
+}
+
+extension ExpectedTokenField: Decodable {
+    init(from decoder: any Decoder) throws {
+        let container = try decoder.singleValueContainer()
+        if let s = try? container.decode(String.self) {
+            self = .str(s)
+        } else if let b = try? container.decode(Bool.self) {
+            self = .bool(b)
+        } else if let d = try? container.decode([String: String].self) {
```

Comment:
Swift's `Dictionary` is unordered, so this line makes unit tests fail randomly.

Reply:
Can I decode a JSON object `{}` as an ordered dictionary or an array?

Reply:
It seems that `Codable`'s keyed containers do not guarantee that they preserve the ordering of the items they contain.

https://github.com/apple/swift-collections/blob/939cfd25234472b4dc91c3caeab304d15bca9a73/Sources/SortedCollections/SortedDictionary/SortedDictionary%2BCodable.swift#L23

Reply:
According to https://html.spec.whatwg.org/multipage/parsing.html#overview-of-the-parsing-model:

> While the exact order of attributes is [implementation-defined](https://infra.spec.whatwg.org/#implementation-defined), and may depend on factors such as the order that the attributes were given in the original markup, the sort order must be stable, such that consecutive invocations of this algorithm serialize an element's attributes in the same order.

So I decided to use unordered something to store attributes in `Tokenizer` and here.